### PR TITLE
Adding Websockets

### PR DESCRIPTION
**Pull request by jhjaggars**
This PR introduces WebSocket support for connections.

Currently when you start a node in listening mode a WebSocket server is also started on port+1.  Peers can be added by appending ws[s]:// to the front of the address.

### CODE REVIEW

**Comment by psav** [comment]
Remember to add aiohttp to the requirements file/setup.py

File: receptor/__main__.py [hunk 1]
```diff
@@ -42,8 +42,8 @@ def main(args=None):
 
     try:
         config.go()
-    except Exception as e:
-        logger.error("An error occured while running receptor:\n%s" % (str(e),))
+    except Exception:
+        logger.exception("main: an error occured while running receptor")
```
Comment (matburt) [inline receptor/__main__.py:46]: should we hide the reason?

File: receptor/messages/envelope.py [hunk 2]
```diff
@@ -99,6 +102,7 @@ def __init__(self, loop=None):
             await self.handle_frame(rest)
 
     async def finish(self):
+        logger.debug("finish: %s", self.current_frame)
```
Comment (matburt) [inline receptor/messages/envelope.py]: I will say, debug mode is getting a little.... verbose ;)

File: receptor/node.py [hunk 3]
```diff
@@ -15,8 +21,23 @@ def mainloop(receptor, ping_interval=None, loop=asyncio.get_event_loop(), skip_r
...
     for peer in config.node_peers:
-        loop.create_task(create_peer(receptor, loop, *peer.strip().split(":", 1)))
+        parsed = parse_peer(peer)
+        if parsed.scheme == "receptor":
+            loop.create_task(create_peer(receptor, loop, parsed.hostname, parsed.port))
+        elif parsed.scheme in ("ws", "wss"):
+            c = WSClient(receptor, loop)
+            loop.create_task(c.connect(peer))
+        else:
+            print(f"invalid peer: {peer} -> {parsed}")
```
Comment (matburt) [inline receptor/node.py]: maybe `logger.warn`?

**Review by matburt — APPROVED** [review]
This is really just some logging nitpicking.

File: receptor/__main__.py (hunk 1 above)
Comment (jhjaggars) [reply receptor/__main__.py:46]: The reason will be in the stack trace (which I found more helpful), but it is longer.

File: receptor/messages/envelope.py (hunk 2 above)
Comment (jhjaggars) [reply receptor/messages/envelope.py]: Yeah, I agree.

File: receptor/ws.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    async def _wait_handshake(self, ws):
+        logger.debug("serve: waiting for HI")
+        response = await self.buf.get()  # TODO: deal with timeout
+        self.remote_id = response.header["id"]
+        self.register(ws)
+
+
+class WSClient(WSBase):
+    async def connect(self, uri):
+        async with aiohttp.ClientSession().ws_connect(uri) as ws:
```
Comment (chrismeyersfsu) [inline receptor/ws.py:92]: Tower just started using `aiohttp`. You probably want to get a handle on that session and call `await session.close()` or use a context manager to handle the closing or prepared to be peppered with logger messages `Unclosed client session`

File: receptor/ws.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                await self._wait_handshake(ws)
+                await self.start_processing(ws)
+                logger.debug("connect: normal exit")
+            except Exception:
+                logger.exception("connect")
+            finally:
+                self.unregister(ws)
+                await asyncio.sleep(5)
+                logger.debug("connect: reconnecting")
+                self.loop.create_task(self.connect(uri))
```
Comment (chrismeyersfsu) [inline receptor/ws.py:105]: I totally stole this recursive reconnect hotness.

**Pull request merged**